[clang][UBSan] Add implicit conversion check for bitfields

Zonotora opened the pull request:
This patch implements the implicit truncation and implicit sign change checks for bitfields using UBSan. E.g.,
`-fsanitize=implicit-bitfield-truncation` and `-fsanitize=implicit-bitfield-sign-change`. 
~~However, right now some unnecessary emits are generated that ideally will be removed in the future. Let me explain.~~

~~This commit implements the truncation and sign change checks whenever we are emitting a `EmitStoreThroughBitfieldLValue`. I use the `llvm::Value` updated through `EmitStoreThroughBitfieldLValue` to check whether the value changed or not.  However, calling `EmitStoreThroughBitfieldLValue` means we have already evaluated RHS and potentially emitted a truncation or sign check already, which is the case if we have a bitfield of type `char` and the RHS is of type `int`. Thus, in this case we may report the truncation or sign change error two times at runtime under the right circumstances which is not ideal. E.g.,~~

```c++
#include <stdio.h>
#include <math.h>

typedef struct {
    unsigned char a:4;
} X;

int main(void) {
    X x;
    unsigned int a = 272;
    x.a = a;
    return 0;
}
```
~~produces~~
```bash
main.c:11:11: runtime error: implicit conversion from type 'unsigned int' of value 272 (32-bit, unsigned) to type 'unsigned char' changed the value to 16 (8-bit, unsigned)
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior main.c:12:11 in
main.c:11:9: runtime error: implicit conversion from type 'unsigned int' of value 16 (32-bit, unsigned) to type 'unsigned char' changed the value to 0 (4-bit bitfield, unsigned)
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior main.c:12:9 in
```
~~because we truncate the value 272 into a char when evaluating the RHS (resulting in the value 16) and then truncate it again when assigning it to the bitfield.~~

~~I thought of disabling the relevant sanitizer checks when evaluating the RHS, but to make it work for all cases I need to extract the previous (before casting it to char in the above case) `llvm::Value` from `Visit(E->getRHS())` in order to compare it, which seemed more complicated for a first approach. The current approach is easier to iterate upon and is more isolated, but again, not ideal.~~

I would appreciate feedback so that I can advance this to a mergeable state. This is my first commit here, so I have probably missed some essentials!

Zonotora commented:
Hi, @vitalybuka @zygoloid @AaronBallman you might have something to say about this commit

vitalybuka commented:
Is is UB?

vitalybuka commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1094,6 +1114,27 @@ void ScalarExprEmitter::EmitIntegerTruncationCheck(Value *Src, QualType SrcType,
                 {Src, Dst});
 }
 
+static llvm::Value *EmitIsNegativeTestHelper(Value *V, QualType VType,
```
Please extract lambda to function into a separate patch

Zonotora replied:
fixed

vitalybuka replied:
I still see it here, does it needs rebase, or that patch needs to be landed?

vitalybuka commented:
@LebedevRI who added implicit-integer-truncation

vitalybuka commented:
> Is is UB?

I guess it follows existing convention `implicit-integer-truncation` is convenience check and `implicit-signed-integer-truncation` is UB

Zonotora committed:
[clang] Extract negativity check lambda to function

Zonotora committed:
[clang][UBSan] Add implicit conversion check for bitfields

This patch implements the implicit truncation and implicit
sign change checks for bitfields.

Zonotora commented:
> Is is UB?

Don't think so? To quote [1]:
> Issues caught by these sanitizers are not undefined behavior, but are often unintentional.

where "these sanitizers" refer to `implicit-unsigned-integer-truncation`, `implicit-signed-integer-truncation` and `implicit-integer-sign-change`.

[1] https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

Zonotora commented:
Feedback @vitalybuka @AaronBallman @LebedevRI? :smiley:

efriedma-quic commented:
Please add test coverage for compound assignment and increment/decrement.

This seems like a reasonable extension of the existing integer truncation checks, but we might want to consider giving it a unique name.  Otherwise, people using integer truncation checks will have trouble upgrading their compiler.

Zonotora commented:
> Please add test coverage for compound assignment and increment/decrement.
> 
> This seems like a reasonable extension of the existing integer truncation checks, but we might want to consider giving it a unique name. Otherwise, people using integer truncation checks will have trouble upgrading their compiler.

Sure 👍 Fair enough, what name do you suggest? Replace "integer" with "bitfield" as in `implicit-unsigned-bitfield-truncation`, `implicit-signed-bitfield-truncation` and `implicit-bitfield-sign-change`? Or just have a single flag `implicit-bitfield-conversion` for now?

vabridgers commented:
LGTM, but someone else must approve. Thanks for doing this work, Axel!

AaronBallman commented:
Precommit CI seems to have found some related failures that should be investigated. I'd love to hear opinions from the codegen code owners on this, as well as @zygoloid as UBSan code owner. I'm not opposed to the changes, though I do find it somewhat strange to add them to UBSan given that they're not undefined behavior (even though there's precedent).

rjmccall commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -122,6 +122,26 @@ Non-comprehensive list of changes in this release
+- ``-fsanitize=implicit-integer-conversion`` a group to include
+  ``-fsanitize=implicit-unsigned-integer-truncation``,
+  ``-fsanitize=implicit-signed-integer-truncation`` and
+  ``implicit-integer-sign-change``.
```
Do we really need a million different flags here?  The reason we have different checks for different signednesses with the normal conversion checks, I think, is because the standard actually treats them differently: some of those checks are actually checking for UB.  Since there's no UB here in any case, I think you may be overthinking this.  Consider just having `-fsanitize=implicit-bitfield-conversion`, which covers every case in which the value we'd read out the bitfield is not the value that we assigned into it.

Zonotora replied:
This was my first thought as well, it just seemed intuitive to use the same flags as for integer conversion. I think all checks included in `implicit-conversion` are not UB. Anyway, I will revert to only having `-fsanitize=implicit-bitfield-conversion`!

rjmccall commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5570,11 +5570,44 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+    // If LHS refers to a bitfield we want to retrieve the value before
+    // implicit conversion between the bitfield type and the RHS type
+    // and evaluate RHS without integer sanitizer checks (if passed)
+    if (auto *ICE = RetrieveImplicitCastExprForBitfieldSanitizer(E)) {
```
When we say "conversion between A and B", it typically means A is being converted to B, but I think you're actually talking about the reverse here.

Please pull the basic conditions for whether to emit this check into this function (LHS is a bit-field + the sanitizer option); it will make the data flow much clearer.  You can just evaluate them once and then check it again below.

Please test for a bit-field LHS before checking for the sanitizer option.

The function call here should be something like `getOriginalRHSForBitfieldSanitizer`.

Zonotora replied:
Sounds good!

rjmccall commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5570,11 +5570,44 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+    // If sanitizer checks are not used, we do not use the result
+    // Hence, use EmitStoreThroughLValue instead
+    if (SanOpts.hasOneOf(SanitizerKind::ImplicitBitfieldConversion) &&
+        LV.isBitField() && RV.isScalar()) {
```
Integer types always use scalar emission.

rjmccall commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -324,6 +326,19 @@ class ScalarExprEmitter
                                   QualType DstType, SourceLocation Loc);
 
+  /// Emit a check that an [implicit] truncation of a bitfield does not
+  /// discard any bits. It is not UB, so we use the value after truncation.
```
The last sentence of the comment seems out of place here (and on the next function).

rjmccall commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1035,7 +1050,7 @@ EmitIntegerTruncationCheckHelper(Value *Src, QualType SrcType, Value *Dst,
 
   llvm::Value *Check = nullptr;
-  // 1. Extend the truncated value back to the same width as the Src.
+  // 1. Convert the Dst back to the same type as Src.
```
This is not an improvement in the comment.

Zonotora replied:
You are correct 😃

rjmccall commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1097,6 +1112,27 @@ void ScalarExprEmitter::EmitIntegerTruncationCheck(Value *Src, QualType SrcType,
+static llvm::Value *EmitIsNegativeTestHelper(Value *V, QualType VType,
+                                             const char *Name,
+                                             CGBuilderTy &Builder) {
+  // NOTE: zero value is considered to be non-negative.
```
This is self-evident and does not need to be in a comment.  Actually, most of the comments in this function are pretty unnecessary; you can just say something like "If the type is unsigned, the value is never negative." and leave everything else self-documenting.

Zonotora replied:
Leftover comment from previous patches, will update.

rjmccall commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1097,6 +1112,27 @@ void ScalarExprEmitter::EmitIntegerTruncationCheck(Value *Src, QualType SrcType,
+  llvm::Type *VTy = V->getType();
+  if (!VSigned) {
+    // If the value is unsigned, then it is never negative.
+    // FIXME: can we encounter non-scalar VTy here?
```
No, you cannot.

Zonotora replied:
Leftover comment from previous patches, will update.

rjmccall commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -1239,6 +1257,228 @@ void ScalarExprEmitter::EmitIntegerSignChangeCheck(Value *Src, QualType SrcType,
+  llvm::Type *SrcTy = Src->getType();
+  llvm::Type *DstTy = Dst->getType();
+  (void)SrcTy; // Only used in assert()
+  (void)DstTy; // Only used in assert()
```
Then you should surround these variables with `#ifndef NDEBUG`.  But mostly you don't need to do this; `CreateIntCast` will assert this for you.

rjmccall reviewed (commented):
You should handle compound assignments.

Zonotora committed:
[clang][UBSan] Update testcases

Update testcases due to new -fsanitize options regarding bitfields

Zonotora commented:
@rjmccall fixed feedback

> You should handle compound assignments.

I believe I already support compound assignments (and PrePostIncDec), if I am not missing something? :smiley:

rjmccall commented:
> @rjmccall fixed feedback
> 
> > You should handle compound assignments.
> 
> I believe I already support compound assignments (and PrePostIncDec), if I am not missing something? 😃

Ah, maybe you are, sorry.  I was expecting it in `CGExpr.cpp` and forgot that that code was actually in `CGExprScalar.cpp`.

Zonotora commented:
@zygoloid what do you think?

Zonotora commented:
> Precommit CI seems to have found some related failures that should be investigated. I'd love to hear opinions from the codegen code owners on this, as well as @zygoloid as UBSan code owner. I'm not opposed to the changes, though I do find it somewhat strange to add them to UBSan given that they're not undefined behavior (even though there's precedent).

@AaronBallman CI Linux is successful, but "Trigger build on clang-ci" seems to be failing on unrelated format and windows build? I also find it strange that these checks are included in UBSan. However, it might be better to refactor this in a future PR, as you say, there are reasons to include it in UBSan now because of the current `-fsanitize=implicit-conversion` flag. How do we move this forward and merge the PR?

Zonotora commented:
@zygoloid @AaronBallman any updates?

zygoloid commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,50 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+    llvm::Value *Previous = nullptr;
+    RValue RV;
+    QualType SrcType = E->getRHS()->getType();
+    // Check if LHS is a bitfield and sanitizer checks are enabled
```
(nit, throughout) Coding style is to end comments with a period.

zygoloid commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,50 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+    QualType SrcType = E->getRHS()->getType();
+    // Check if LHS is a bitfield and sanitizer checks are enabled
+    if (E->getLHS()->refersToBitField() &&
+        SanOpts.hasOneOf(SanitizerKind::ImplicitConversion |
+                         SanitizerKind::ImplicitBitfieldConversion)) {
```
```suggestion
        SanOpts.hasOneOf(SanitizerKind::ImplicitBitfieldConversion)) {
```
No need to check for the containing group; that should be handled automatically.

zygoloid commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,50 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+      }
+    }
+
+    // Otherwise, visit RHS as usual
+    if (!Previous)
+      RV = EmitAnyExpr(E->getRHS());
```
Would it be correct to do the above, more-elaborate lowering instead of this regardless of whether a sanitizer is enabled? I'd have more confidence that we're performing a correct emission if we used the same code regardless of sanitizer mode.

zygoloid commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,50 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
       EmitNullabilityCheck(LV, RV.getScalarVal(), E->getExprLoc());
-    EmitStoreThroughLValue(RV, LV);
+
+    // Passing a pointer EmitStoreThroughBitfieldLValue will emit the result
+    // If sanitizer checks are not used, we do not use the result
+    // Hence, use EmitStoreThroughLValue instead
+    if (LV.isBitField() && RV.isScalar() &&
+        SanOpts.hasOneOf(SanitizerKind::ImplicitBitfieldConversion)) {
+      // If the expression contained an implicit conversion, make sure
+      // to use the value before the scalar conversion.
+      llvm::Value *Src = Previous ? Previous : RV.getScalarVal();
+      QualType DstType = E->getLHS()->getType();
+      llvm::Value *Result;
+      EmitStoreThroughBitfieldLValue(RV, LV, &Result);
+      EmitBitfieldConversionCheck(Src, SrcType, Result, DstType,
+                                  LV.getBitFieldInfo(), E->getExprLoc());
+    } else
+      EmitStoreThroughLValue(RV, LV);
```
Same here: can we use the same code regardless, and just make a choice between passing in `&Result` and passing in `nullptr` to `EmitStoreThroughBitfieldLValue`, and a choice as to whether we `EmitBitfieldConversionCheck`, based on the sanitizer mode?

zygoloid commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -4564,15 +4737,41 @@ Value *ScalarExprEmitter::VisitBinAssign(const BinaryOperator *E) {
+    QualType SrcType = E->getRHS()->getType();
+    // Check if LHS is a bitfield and sanitizer checks are enabled
+    if (E->getLHS()->refersToBitField() &&
+        CGF.SanOpts.hasOneOf(SanitizerKind::ImplicitConversion |
+                             SanitizerKind::ImplicitBitfieldConversion)) {
```
```suggestion
        CGF.SanOpts.hasOneOf(SanitizerKind::ImplicitBitfieldConversion)) {
```

zygoloid reviewed (commented):
Generally this looks good to me, thanks.

zygoloid commented on compiler-rt/lib/ubsan/ubsan_handlers.cpp:
```diff
@@ -555,13 +555,11 @@ static void handleImplicitConversion(ImplicitConversionData *Data,
-
   bool SrcSigned = SrcTy.isSignedIntegerTy();
   bool DstSigned = DstTy.isSignedIntegerTy();
+  ErrorType ET = ErrorType::GenericUB;
```
This seems suspect, given that this handler is detecting things that aren't UB.

Zonotora replied:
I have only moved this line... I think changing the error type is for a future PR as you comment further down.

zygoloid replied:
Ah, OK. That makes sense.

zygoloid commented:
> I'm not opposed to the changes, though I do find it somewhat strange to add them to UBSan given that they're not undefined behavior (even though there's precedent).

This is adding checks to `-fsanitize=implicit-conversion`, which is not part of `-fsanitize=undefined`. It's sort of ambiguous whether UBSan means `-fsanitize=undefined` (that is, checks for UB that can be done without a big runtime), or whether it means any of the checks that the UBSan infrastructure and runtime power, but the established precedent is we also use the UBSan infrastructure to do non-UB data-loss checks like this, and this patch seems like a logical extension of the existing behavior.

I'd be happier if we drew a brighter distinction between the UB checks and the non-UB checks that our sanitizers perform. For example, perhaps the non-UB sanitizers shouldn't be listed in https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html but in a different page, or perhaps at least in a different section from the other checks. But I think that's independent of this PR, and in the context of this PR, I don't think we should be re-litigating whether this kind of data-loss-through-conversion check is in scope.

Zonotora commented:
@zygoloid Thanks for your feedback! Fixed all comments except the last one. I totally agree with you about the distinction of UB checks, and that should be part of a different future PR!

zygoloid commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,52 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
+    // is enabled) use it to check for an implicit conversion.
+    if (E->getLHS()->refersToBitField()) {
+      // Get the RHS before scalar conversion.
+      if (auto *ICE = GetOriginalRHSForBitfieldSanitizer(E)) {
```
What do you think about renaming this to `GetOriginalRHSForBitfieldAssignment`, given that we now use it unconditionally for all lowering of bitfield assignments?

Zonotora replied:
Will do!

zygoloid reviewed (commented):
Looks good to me. @AaronBallman Did you have remaining concerns here?

AaronBallman reviewed (approved):
LGTM, so long as @rjmccall or @efriedma-quic don't have concerns about codegen changes. Please give them a few days to chime in before landing.

efriedma-quic commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -5571,11 +5571,52 @@ LValue CodeGenFunction::EmitBinaryOperatorLValue(const BinaryOperator *E) {
     }
 
-    RValue RV = EmitAnyExpr(E->getRHS());
+    llvm::Value *Previous = nullptr;
```
Can we de-duplicate this code with the corresponding code in CGExprScalar, similar to the way EmitCompoundAssignmentLValue works?

Zonotora replied:
I would also like to have it work similar to how EmitCompoundAssignmentLValue works, however it doesn't seem completely trivial to make this change? Maybe it is better suited in its own PR as well? But I can definitely convert:
```cpp
if (E->getLHS()->refersToBitField()) {
      // Get the RHS before scalar conversion.
  if (auto *ICE = GetOriginalRHSForBitfieldAssignment(E)) {
    SrcType = ICE->getSubExpr()->getType();
    Previous = EmitScalarExpr(ICE->getSubExpr());
    // Pass default ScalarConversionOpts to avoid emitting
    // integer sanitizer checks as E refers to bitfield.
    llvm::Value *RHS = EmitScalarConversion(
        Previous, SrcType, ICE->getType(), ICE->getExprLoc());
    RV = RValue::get(RHS);
  }
}
```
to something like:
```cpp
if (E->getLHS()->refersToBitField())
  Previous = EmitForOriginalRHSBitfieldAssignment(E, &RHS));
```
to at least de-duplicate some parts of it...

efriedma-quic replied:
Sure, if you want to leave the larger refactoring for later, that's fine.

efriedma-quic reviewed (commented):
I'm a bit concerned about the lack of C++ testcases, since you're making changes to some C++-only codepaths.

efriedma-quic commented on clang/test/CodeGen/ubsan-bitfield-conversion.c:
```diff
@@ -0,0 +1,61 @@
+// RUN: %clang -fsanitize=implicit-bitfield-conversion -target x86_64-linux -S -emit-llvm -o - %s | FileCheck %s --check-prefixes=CHECK,CHECK-BITFIELD-CONVERSION
```
I'd like to see some testcases that run in C++ mode.

Zonotora replied:
I will add some testcases later today!

Zonotora replied:
Added!

Zonotora commented on compiler-rt/lib/ubsan/ubsan_handlers.h:
```diff
@@ -147,6 +147,7 @@ struct ImplicitConversionData {
   const TypeDescriptor &FromType;
   const TypeDescriptor &ToType;
   /* ImplicitConversionCheckKind */ unsigned char Kind;
+  unsigned int BitfieldBits;
```
@zygoloid

AaronBallman replied:
Ping @zygoloid -- I'm not certain if we have an ABI stability policy for sanitizers or not.

Zonotora commented:
@LebedevRI LGTM?

Zonotora commented:
Also, I don't have commit access (first PR here), someone has to commit on my behalf!

Zonotora commented:
> Also, I don't have commit access (first PR here), someone has to commit on my behalf!

@AaronBallman

AaronBallman commented:
> > Also, I don't have commit access (first PR here), someone has to commit on my behalf!
> 
> @AaronBallman

Thank you for mentioning this! I'll commit on your behalf, but I'd like to hear from @zygoloid on the ABI question. It would also be nice if @efriedma-quic signed off on the testing changes (I think they're sufficient, but he may have other cases in mind).

zygoloid commented:
I don't think we've established an explicit policy, and we've made ABI-breaking changes previously. I think we should avoid breaks within a major release version (don't backport this) to avoid giving packagers headaches: the ubsan runtime is installed in a versioned directory, but I don't think that includes the patch version.

Having an ABI break here seems OK and in line with historic precedent. (But we ought to document our ABI policy for our runtime libraries.)

efriedma-quic reviewed (approved):
LGTM

Zonotora commented:
@AaronBallman looks good?

vitalybuka commented:
This bot started inconsistently crash https://lab.llvm.org/buildbot/#/builders/75/builds/45086
Few builds before are green and tests only, so I suspect this patch.

vitalybuka commented:
These are consistently broken after the patch:
https://lab.llvm.org/buildbot/#/builders/77/builds/36021
https://lab.llvm.org/buildbot/#/builders/37/builds/33244
https://lab.llvm.org/buildbot/#/builders/19/builds/26087

AaronBallman commented:
Thank you for catching that and doing the revert!

vitalybuka commented:
Probably  1189e87951e59a81ee097eae847c06008276fef1 fixed the issue.
I will revert my revert.
